### PR TITLE
Task #1178 problem with Commit Synchronization in Collaborative Environment

### PR DESCRIPTION
**Pull request by DimitriDiantos**
To resolve this, I suggest reviewing the code behind the "commit project to repository" button, ensuring that it performs an update before each commit.

### CODE REVIEW

**Comment by franzTobiasDLR** [comment]
If your not yet done with your feature, please add the "Draft:" thing to the title of the PR, so that it is clear that it should not yet be merged... And only add us as reviers once our done ;) Otherwise we don't know when do look at the changes... ;)

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java [hunk 1]
```diff
@@ -35,8 +37,28 @@ protected IVirSatVersionControlBackend createVersionControlBackend() {
...
+					new PullResultDialog(Display.getDefault().getActiveShell(), gitRepository, pullResult);
+			pullResultDialog.open();
+		});
+	}
+	
 	@Override
 	protected void doCommit(IProject project, String message, IProgressMonitor monitor) throws Exception {
+		
+        // Update before committing
+        updateBeforeCommit(project, monitor);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java:61]: I think if you implement it here, it only works for git not SVN... Maybe move the implementation to the AVersionControlCommitHandler?

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitUpdateHandler.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ protected void doUpdate(IProject project, IProgressMonitor monitor) throws Excep
 		
 		Repository gitRepository = RepositoryMapping.getMapping(project).getRepository();
 
-		// Build the dialog in the UI thread
+		// Build the dialog in the UI threads
```
Comment (dellerDLR) [inline de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitUpdateHandler.java:46]: I think the comment is right, there is just one UI thread :)

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java (hunk 1 above)
Comment (dellerDLR) [reply de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java:61]: I agree to that

**Review by dellerDLR — CHANGES_REQUESTED** [review]

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitUpdateHandler.java (hunk 2 above)
Comment (DimitriDiantos) [reply de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitUpdateHandler.java:46]: Yes, i think too.

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java [hunk 3]
```diff
@@ -27,34 +27,35 @@
 /**
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java:27]: Maybe you can reset this file as you didn't do any changes to it?

File: de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java [hunk 4]
```diff
@@ -135,21 +135,28 @@ protected void testCommitProjectAssert() {
...
+	    historyView.show();
+	    SWTBotTable historyTable =  historyView.bot().table();
+	    SWTBotTableItem historyTableItem0 = historyTable.getTableItem(0);
+
+	    // Retrieve the commit message
+	    String commitMessage = historyTableItem0.getText(1);
+	    
+	    // If the commit message contains the phrase "Backend Local Commit Before Pull", extract the actual commit message
+	    if (commitMessage.contains("Backend Local Commit Before Pull")) {
+	        commitMessage = commitMessage.replace("Backend Local Commit Before Pull: ", "");
```
[the remark this answers is not in the document]
Comment (DimitriDiantos) [reply de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java:153]: That is right, i removed it.

File: de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java (hunk 3 above)
Comment (DimitriDiantos) [reply de.dlr.sc.virsat.team.ui/src/de/dlr/sc/virsat/team/ui/handler/git/GitCommitHandler.java:27]: I did it :) .

File: de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dellerDLR) [reply de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java]: Instead of changing (replacing "Backend Local Commit Before Pull: ") the string inside the test, I think the cleaner version would be to change the string in the assert function in the test

File: de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java
Comment (dellerDLR) [inline de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java:1]: I think the changes here can be reverted completly and just the assert function in the class where this function is called has to be changed (See other comment).

**Review by dellerDLR — CHANGES_REQUESTED** [review]

File: de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java
Comment (dellerDLR) [reply de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java:1]: @franzTobiasDLR  What do you think?

Comment (dellerDLR) [inline de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java:1]: Can you revert the changes to the state of the commit, before you started to change something?
By doing so, it wont appear in the list of changed files.
Other than that it is ready to be merged by my side :)

**Review by dellerDLR — CHANGES_REQUESTED** [review]

File: de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java
Comment (DimitriDiantos) [reply de.dlr.sc.virsat.swtbot.test/src/de/dlr/sc/virsat/swtbot/test/versioningbackend/GitVersioningBackendTest.java:1]: Alright. I did it.

**Review by franzTobiasDLR — APPROVED** [review]
